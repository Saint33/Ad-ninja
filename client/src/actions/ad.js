import axios from 'axios';

export const deleteAd = id => {
    const request = axios.delete(`/api/ad?id=${id}`)
        .then(response => response.data)
    return {
        type: 'DELETE_AD',
        payload: request
    }
}

export const deleteFromFavorites = data => {
    const request = axios.post(`/api/ad/delete-from-favorites`, data)
        .then(response => {console.log(response, data)
            return response.data    
        })
    return {
        type: 'DELETE_AD_FROM_FAVORITES',
        payload: request
    }
}

export const getAd = id => {
    const request = axios.get(`/api/ad?id=${id}`)
        
    return (dispatch) => {
        request.then(({data}) => {
            let ad = data;
            axios.get(`/api/user/${ad.ownerId}`)
                .then(({data}) => {
                    let adInfo = {
                        ...ad,
                        owner: data
                    }
                    dispatch({
                        type: 'GET_AD',
                        payload: adInfo
                    })
                })
        })
    }
}

export const userAds = id => {
    const request = axios.get(`/api/ad/user-ads?id=${id}`)
        .then(response => response.data.docs)

    return {
        type: 'USER_ADS',
        payload: request
    }
}

export const addToFavorites = data => {
    const request = axios.post('/api/ad/add-to-favorites', data)
        .then(response => console.log(response))

    return {
        type: 'ADD_TO_FAVORITES',
        payload: request
    }
}
