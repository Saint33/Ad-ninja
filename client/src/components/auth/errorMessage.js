import React from 'react'
import { connect } from 'react-redux'
import { Alert } from 'reactstrap'
import {Animated} from "react-animated-css";

let ErrorMessage = (props) => {
	return (
        props.errorMessage ?
            <Animated animationIn="flash" animationOut="fadeOut" isVisible={true}>
                <Alert className="error" color="danger">
                    {props.errorMessage}
                </Alert> 
            </Animated>
        : null
	);
};


const mapStateToProps = (state) => ({
	errorMessage: state.user.login.errorMessage
});

ErrorMessage = connect(mapStateToProps)(ErrorMessage);

export default ErrorMessage;