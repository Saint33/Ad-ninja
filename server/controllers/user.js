const { User } = require('../models/user');

exports.getUser = (req, res, next) => {
    let id = req.params.id;

    User.findById(id, (err, user) => {
        if(err) return res.status(400).send(err);
        res.send(user);
    })
};

exports.getUsers = (req, res, next) => {
    User.find({}, (err, user) => {
        if(err) return res.status(400).send(err);
        res.status(200).send(user);
    });
};

exports.getUserInfo = (req, res, next) => {
    let id = req.query.id;

    User.findById(id, (err, doc) => {
        if(err) return res.status(400).send(err);
        res.send({
            username: doc.username,
            id: doc._id
        });
    });
};