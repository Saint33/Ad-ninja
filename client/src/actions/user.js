import axios from 'axios';
import { userAds } from './ad';

export const login = ({email, password}) => {
    const request = axios.post('/api/auth/login', {email, password})
        .then(response => response.data)
    return{
        type: 'USER_LOGIN',
        payload: request
    }
}

export const auth = () => {
    const request = axios.get('/api/auth')
        .then(response => response.data)

    return {
        type: 'USER_AUTH',
        payload: request
    }
}

export const logout = () => {
    const request = axios.get('/api/auth/logout')
        .then(response => response.data)

    return {
        type: 'USER_LOGOUT',
        payload: request
    }
}

export const getUser = (id) => {
    const request = axios.get(`/api/user/${id}`)
        .then(response => response.data)

    return {
        type: 'GET_USER',
        payload: request
    }  
}

export const getUserWithAds = (id) => {
    return dispatch => {
        dispatch(getUser(id))
            .then(() => dispatch(userAds(id)))
    }

}