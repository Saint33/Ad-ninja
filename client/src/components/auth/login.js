import React, { Component } from 'react';
import { InputGroup, Input } from 'reactstrap';
import Button from '../UI/button';
import { connect } from 'react-redux';
import { login } from '../../actions/user';
import { withRouter } from 'react-router-dom';
import ErrorMessage from './errorMessage';

export class Login extends Component {
    state = { 
        email: '',
        password: '',
        error: ''
    }
    
    handleInputEmail = (e) => {
        this.setState({email: e.target.value})
    }

    handleInputPassword = (e) => {
        this.setState({password: e.target.value})
    }

    componentWillReceiveProps(nextProps){
        if(nextProps.user.login.isAuth){
            this.props.closeModal();
            this.props.history.push('/');
        }
    }

    submitForm = (e) => {
        e.preventDefault();
        let loginData = {email:this.state.email, password: this.state.password}
        this.props.dispatch(login(loginData));
    }

    render() {
        return (
                <form className="login" onSubmit={this.submitForm} >
                    <InputGroup>
                        <Input 
                            type="email"
                            placeholder="Электронная почта"
                            value={this.state.email}
                            onChange={this.handleInputEmail}
                            className="login__input"
                            name="email"
                        />
                            <br />
                    </InputGroup>
                    <InputGroup>
                    <Input 
                            type="password"
                            placeholder="Пароль"
                            value={this.state.password}
                            onChange={this.handleInputPassword}
                            className="login__input"
                            name="password"
                        />
                    </InputGroup>
                    <div className="button_wrapper" key="button">
                        <Button   key="buttondassd"
                            type="submit"
                        >Войти</Button>
                    </div>
                    
                        <ErrorMessage />

                </form>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        user: state.user
    }
}

export default connect(mapStateToProps)(withRouter(Login));