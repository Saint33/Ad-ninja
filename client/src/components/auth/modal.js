import React, { Component } from 'react';
import Modal from 'react-modal';
import Login from '../auth/login';
import Register from '../auth/register';
import { connect } from 'react-redux';
import { openModal, closeModal } from '../../actions/user';
import { Motion, spring, TransitionMotion } from 'react-motion';

const customStyles = {
    content : {
      top                   : '50%',
      left                  : '50%',
      right                 : 'auto',
      bottom                : 'auto',
      marginRight           : '-50%',
      transform             : 'translate(-50%, -50%)',
      borderRadius         : '10px'
    }
  };

class LoginRegisterModal extends Component {
    constructor(props) {
        super(props);
    
        this.state = {
          login: true,
          register: false
        };
    }

    handleLogin = () => {
        this.setState((prevState) => ({...prevState, login: true, register: false}))
    }

    handleRegistration = () => {
        this.setState((prevState) => ({...prevState, login: false, register: true}))
    }
    handleOpenModal = () => {
        this.props.dispatch(openModal())
    }
    
    handleCloseModal = () => {
        this.props.dispatch(closeModal())
    }
    willLeave() {
        // triggered when c's gone. Keeping c until its width/height reach 0.
        return {width: spring(0), height: spring(0), opacity: spring(0)};
      }
    render() {
        return (
            <Modal
                isOpen={this.props.modalIsOpen}
                onAfterOpen={this.afterOpenModal}
                onRequestClose={this.handleCloseModal}
                style={customStyles}
                contentLabel="Example Modal"
            >   
            <Motion
                defaultStyle={{opacity: 0, x: 0, rotate: 0}}
                style={{
                    opacity: spring(1, {stiffness: 110, damping: 37}),
                    x: spring(this.state.login ? 0 : 115),
                    rotate: spring(this.state.login? -3 : 3)
                }}
            >
                { interpolatingStyle => 
                    <img 
                        style={{
                            opacity: interpolatingStyle.opacity,
                            transform: `translateX(${interpolatingStyle.x}px) rotate(${interpolatingStyle.rotate}turn)`
                        }}
                        src="/images/suricane.png" 
                        alt="suricane" 
                        className="modal-suricane"
                    />
                }
            </Motion>
            
            <h2 className="login__title">
                <span 
                    className={this.state.login ? 'modal-title-active modal-title': 'modal-title'} 
                    onClick={this.handleLogin}
                    >Вход</span> / 
                <span 
                    className={this.state.register ? 'modal-title-active modal-title' : 'modal-title'} 
                    onClick={this.handleRegistration}
                    > Регистрация</span> 
            </h2>
            <div>
                { this.state.login ? 
                    <Login  
                        closeModal={this.handleCloseModal} 
                     />
                    : 
                    <Register  
                         closeModal={this.handleCloseModal} />

                    }
            </div>
            </Modal>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        modalIsOpen: state.user.loginModalIsOpen
    }
}

export default connect(mapStateToProps)(LoginRegisterModal);