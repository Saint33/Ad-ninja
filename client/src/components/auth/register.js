import React, { Component } from 'react';
import { InputGroup, Input } from 'reactstrap';
import { connect } from 'react-redux';
import Button from '../UI/button';
import { register, login } from '../../actions/user';
import { withRouter } from 'react-router-dom';

export class Register extends Component {
    state = { 
        email: '',
        password: '',
        error: '',
        username: '',
        phone: ''
    }
    
    handleInputEmail = (e) => {
        this.setState({email: e.target.value})
    }

    handleInputPassword = (e) => {
        this.setState({password: e.target.value})
    }

    handleInputUsername = (e) => {
        this.setState({username: e.target.value})
    }

    handleInputPhone = (e) => {
        this.setState({phone: e.target.value})
    }

    componentWillReceiveProps(nextProps){
        if(nextProps.user.login.isAuth){
            console.log(nextProps)
            this.props.closeModal();
            this.props.history.push('/');
        }
    }

    submitForm = (e) => {
        e.preventDefault();
        let registrationData = {
            email: this.state.email, 
            password: this.state.password,
            phone: this.state.phone,
            username: this.state.username
        }
        this.props.dispatch(register(registrationData));
    }

    render() {
        return (
                <form className="login" onSubmit={this.submitForm}>
                    <InputGroup>
                        <Input 
                            type="text"
                            placeholder="Имя пользователя"
                            value={this.state.username}
                            onChange={this.handleInputUsername}
                            className="login__input"
                            name="username"
                            />
                            <br />
                    </InputGroup>
                    <InputGroup>
                        <Input 
                            type="text"
                            placeholder="Телефон"
                            value={this.state.phone}
                            onChange={this.handleInputPhone}
                            className="login__input"
                            name="phone"
                            />
                            <br />
                    </InputGroup>
                    <InputGroup>
                        <Input 
                            type="email"
                            placeholder="Электронная почта"
                            value={this.state.email}
                            onChange={this.handleInputEmail}
                            className="login__input"
                            />
                            <br />
                    </InputGroup>
                    <InputGroup>
                    <Input 
                            type="password"
                            placeholder="Пароль"
                            value={this.state.password}
                            onChange={this.handleInputPassword}
                            className="login__input"
                        />
                    </InputGroup>
                    <div className="button_wrapper">
                    <Button 
                        type="submit"
                    >Зарегистрироваться</Button>
                    </div>
                </form>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        user: state.user
    }
}

export default connect(mapStateToProps)(withRouter(Register));