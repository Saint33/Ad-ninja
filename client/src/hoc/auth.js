import React, { Component } from 'react';
import { connect } from 'react-redux';
import { auth } from '../actions/user';
import Spinner from 'react-spinkit';
import { openModal } from '../actions/user';

export default function(ComposedClass, reload){

    class AuthenticationCheck extends Component {
        
        state = {
            loading: true
        }
        
        componentWillMount() {
            this.props.dispatch(auth())
        }

        componentWillReceiveProps(nextProps) {
            this.setState({loading: false})

            if(!nextProps.user.isAuth){
                if(reload === true){
                    // this.props.history.push('/login');  
                    this.props.history.push('/');
                    this.props.dispatch(openModal())                
                }
            } else {
                if(reload === false){
                    this.props.history.push('/');
                }
            }
        }

        render(){
            if(this.state.loading){
                return <Spinner className="loader-position" name='folding-cube' fadeIn="none"/>
            }
            return(
                <ComposedClass {...this.props} user={this.props.user}/>
            )
        }
    }

    const mapStateToProps = (state) => {
        return {
            user: state.user.login
        }
    }

    return connect(mapStateToProps)(AuthenticationCheck);
};
