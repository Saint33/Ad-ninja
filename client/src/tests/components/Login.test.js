
import React from 'react';
import { Login } from '../../components/auth/login';
import { shallow, mount } from 'enzyme';;
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store'
import { login } from '../../actions/user';

describe('Login Component', () => {

    let wrapper;
    const mockDispatch = jest.fn();

    const middlewares = []
    const mockStore = configureStore(middlewares)

    test('should call the mock login function', () => {
        wrapper = shallow(<Login dispatch={mockDispatch}/>)
        wrapper.find('.login').simulate(
            'submit',
            { preventDefault() {}}
        )
        expect(mockDispatch.mock.calls.length).toBe(1)
    })


    test('should be called with the email and password in the state as arguments', () => {
        const fakeUser = { email: 'fakeuser@mail.com', password: 'fakepassword'}
        const initialState = { user: { login: { errorMessage: '' }}}
        const store = mockStore(initialState)
        wrapper = mount(
            <Provider store={store}>
                <Login dispatch={mockDispatch}/>
            </Provider>
        )
        wrapper.find({type: 'email'}).at(0).simulate(
            'change',
            { target:
                { value: fakeUser.email }
            }
        )
        wrapper.find({type: 'password'}).at(0).simulate(
            'change',
            { target:
                { value: fakeUser.password }
            }
        )
        wrapper.find('.login').simulate(
            'submit', 
            {preventDefault() {}}
        )

        expect(wrapper.find('Login').instance().state.email).toEqual(fakeUser.email)
        expect(wrapper.find('Login').instance().state.password).toEqual(fakeUser.password)
        expect(wrapper).toMatchSnapshot();
    })


})
