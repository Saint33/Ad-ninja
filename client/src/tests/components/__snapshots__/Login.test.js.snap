// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should be called with the email and password in the state as arguments 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Login
    dispatch={[Function]}
  >
    <form
      className="login"
      onSubmit={[Function]}
    >
      <InputGroup
        tag="div"
      >
        <div
          className="input-group"
        >
          <Input
            className="login__input"
            name="email"
            onChange={[Function]}
            placeholder="Электронная почта"
            type="email"
            value="fakeuser@mail.com"
          >
            <input
              className="login__input form-control"
              name="email"
              onChange={[Function]}
              placeholder="Электронная почта"
              type="email"
              value="fakeuser@mail.com"
            />
          </Input>
          <br />
        </div>
      </InputGroup>
      <InputGroup
        tag="div"
      >
        <div
          className="input-group"
        >
          <Input
            className="login__input"
            name="password"
            onChange={[Function]}
            placeholder="Пароль"
            type="password"
            value="fakepassword"
          >
            <input
              className="login__input form-control"
              name="password"
              onChange={[Function]}
              placeholder="Пароль"
              type="password"
              value="fakepassword"
            />
          </Input>
        </div>
      </InputGroup>
      <div
        className="button_wrapper"
        key="button"
      >
        <Button
          key="buttondassd"
          type="submit"
        >
          <button
            className="button"
            type="submit"
          >
            Войти
          </button>
        </Button>
      </div>
      <Connect(ErrorMessage)>
        <ErrorMessage
          dispatch={[Function]}
          errorMessage=""
        />
      </Connect(ErrorMessage)>
    </form>
  </Login>
</Provider>
`;
